### DISCUSSION TITLE
Install sqlalchemy via pip failed due to unresolved issue from "greenlet"

### DISCUSSION THREAD

**Discussion by FatalLin** [sqlalchemy/sqlalchemy#9083] [Usage Questions]
Hi team,
our building pipeline is broken and here's the error message we saw.
`#6 259.9     gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -std=c++11 -fPIC -I/usr/local/include/python3.7m -c src/greenlet/greenlet.cpp -o build/temp.linux-x86_64-3.7/src/greenlet/greenlet.o
#6 259.9     src/greenlet/greenlet.cpp:2511:64: error: specialization of ‘template<class IsPy37> const OwnedObject greenlet::Greenlet::context(typename IsPy37::IsIt) const’ in different namespace [-fpermissive]
#6 259.9      Greenlet::context<GREENLET_WHEN_PY37>(GREENLET_WHEN_PY37::Yes) const
#6 259.9                                                                     ^~~~~
#6 259.9     In file included from src/greenlet/greenlet_internal.hpp:20:0,
#6 259.9                      from src/greenlet/greenlet.cpp:19:
#6 259.9     src/greenlet/greenlet_greenlet.hpp:333:27: error:   from definition of ‘template<class IsPy37> const OwnedObject greenlet::Greenlet::context(typename IsPy37::IsIt) const’ [-fpermissive]
#6 259.9              const OwnedObject context(const typename IsPy37::IsIt=nullptr) const;
#6 259.9                                ^~~~~~~
#6 259.9     src/greenlet/greenlet.cpp:2540:71: error: specialization of ‘template<class IsPy37> const OwnedObject greenlet::Greenlet::context(typename IsPy37::IsIt) const’ in different namespace [-fpermissive]
#6 259.9      Greenlet::context<GREENLET_WHEN_NOT_PY37>(GREENLET_WHEN_NOT_PY37::No) const
#6 259.9                                                                            ^~~~~
#6 259.9     In file included from src/greenlet/greenlet_internal.hpp:20:0,
#6 259.9                      from src/greenlet/greenlet.cpp:19:
#6 259.9     src/greenlet/greenlet_greenlet.hpp:333:27: error:   from definition of ‘template<class IsPy37> const OwnedObject greenlet::Greenlet::context(typename IsPy37::IsIt) const’ [-fpermissive]
#6 259.9              const OwnedObject context(const typename IsPy37::IsIt=nullptr) const;
#6 259.9                                ^~~~~~~
#6 259.9     src/greenlet/greenlet.cpp:2562:89: error: specialization of ‘template<class IsPy37> void greenlet::Greenlet::context(greenlet::refs::BorrowedObject, typename IsPy37::IsIt)’ in different namespace [-fpermissive]
#6 259.9      void Greenlet::context<GREENLET_WHEN_PY37>(BorrowedObject given, GREENLET_WHEN_PY37::Yes)
#6 259.9                                                                                              ^
#6 259.9     In file included from src/greenlet/greenlet_internal.hpp:20:0,
#6 259.9                      from src/greenlet/greenlet.cpp:19:
#6 259.9     src/greenlet/greenlet_greenlet.hpp:336:21: error:   from definition of ‘template<class IsPy37> void greenlet::Greenlet::context(greenlet::refs::BorrowedObject, typename IsPy37::IsIt)’ [-fpermissive]
#6 259.9              inline void context(refs::BorrowedObject new_context, typename IsPy37::IsIt=nullptr);
#6 259.9                          ^~~~~~~
#6 259.9     src/greenlet/greenlet.cpp:2597:99: error: specialization of ‘template<class IsPy37> void greenlet::Greenlet::context(greenlet::refs::BorrowedObject, typename IsPy37::IsIt)’ in different namespace [-fpermissive]
#6 259.9      Greenlet::context<GREENLET_WHEN_NOT_PY37>(BorrowedObject UNUSED(given), GREENLET_WHEN_NOT_PY37::No)
#6 259.9                                                                                                        ^
#6 259.9     In file included from src/greenlet/greenlet_internal.hpp:20:0,
#6 259.9                      from src/greenlet/greenlet.cpp:19:
#6 259.9     src/greenlet/greenlet_greenlet.hpp:336:21: error:   from definition of ‘template<class IsPy37> void greenlet::Greenlet::context(greenlet::refs::BorrowedObject, typename IsPy37::IsIt)’ [-fpermissive]
#6 259.9              inline void context(refs::BorrowedObject new_context, typename IsPy37::IsIt=nullptr);
#6 259.9                          ^~~~~~~
#6 259.9     error: command 'gcc' failed with exit status 1
#6 259.9     
#6 259.9     ----------------------------------------
#6 260.2 Command "/usr/local/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-9_euj1gc/greenlet/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-qw3kig0j/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-install-9_euj1gc/greenlet/
`
it seems like there's an unhandled issue on `greenlet` since last Dec.
Our base image python:3.7.0-slim, and both failed while installing sqlalchemy 1.4.11 and 1.4.46.
Do you have any idea how to fix it or work around with it? Thanks.

**Comment by zzzeek** [comment]
you'd ideally want greenlet to be installed from one of its published wheel files: https://pypi.org/project/greenlet/#files 

using standard build tools greenlet is only a dependency if your platform is one of the architectures that they have listed there, so there shouldn't be a circumstance where pip is trying to actually build it from source (which is known to be problematic).  it looks like you're on x86_64 so one of the cp37 wheel files at that link should be getting installed.

right now if you can't install a wheel file and can't build greenlet from source you'd have to comment out greenlet in the setup.cfg file, it's normally an optional dependency.
